### PR TITLE
[SYCL][Driver] Fix fno-gpu-rdc SYCL libraries linking not using opaque pointers

### PR DESCRIPTION
**Pull request by GeorgeWeb**
Fix fno-gpu-rdc SYCL libraries linking not using opaque pointers for non-SPIR targets.

Issue was when specifiying -fno-gpu-rdc for an opaque pointer target, there would be object modules with typed pointers. so llvm-link would need the opaque-pointer flag but that wasn't being added previously due to relying on the default RDC behaviour where the first call to llvm-link linking the SYCL libraries with the kernel module is expected to have the kernel module as the first input. This isn't true in no-rdc mode, where the first input is a SYCL device library, hence checking the RDC mode before setting the state of `LinkSYCLDeviceLibs`.

Test checking the behaviour was added as well.

### CODE REVIEW

File: clang/test/Driver/sycl-no-gpu-rdc-opaque-pointers.cpp
Comment (mdtoguchi) [inline clang/test/Driver/sycl-no-gpu-rdc-opaque-pointers.cpp:15]: There isn't any actual compilation occurring - the presence of any source code is not needed.

Comment (GeorgeWeb) [reply clang/test/Driver/sycl-no-gpu-rdc-opaque-pointers.cpp]: Correct yeah. Thank you! Updated now.

**Comment by sarnex** [comment]
@GeorgeWeb Sorry I was out on a long vacation, back today. I will try to get to this soon, thanks.

File: clang/lib/Driver/ToolChains/SYCL.cpp
Comment (sarnex) [inline clang/lib/Driver/ToolChains/SYCL.cpp:226]: could you put a comment explaining why we need to do this for future reads? thx

File: clang/test/Driver/sycl-no-gpu-rdc-opaque-pointers.cpp
Comment (sarnex) [inline clang/test/Driver/sycl-no-gpu-rdc-opaque-pointers.cpp:8]:
```suggestion
/// FIXME: remove once opaque pointers are supported for all SYCL targets (i.e., SPIR-V)
```

nitpick

**Review by sarnex — APPROVED** [review]
lgtm, the original code is a bit hacky (relying on input file order) but i dont think this change is making it worse

**Comment by GeorgeWeb** [comment]
@bader Is it possible to merge this one? Think it looks good now after Nick's review. Thanks.

**Review by sarnex — APPROVED** [review]

**Comment by bader** [comment]
> @bader Is it possible to merge this one? Think it looks good now after Nick's review. Thanks.

Not yet. The driver team hasn't approved it yet.

Please, add "[SYCL][Driver]" tags to the PR title. Commit message guidelines - https://github.com/intel/llvm/blob/sycl/sycl/doc/developer/ContributeToDPCPP.md#commit-message and https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#pull-request.

Once you get approval from the code owners, the PR can be merged according to our contribution guide. Please, read https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#merge.

**Comment by GeorgeWeb** [comment]
> Once you get approval from the code owners, the PR can be merged according to our contribution guide. Please, read https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#merge.

Got it. Thanks!

**Comment by npmiller** [comment]
ping @mdtoguchi

**Review by mdtoguchi — APPROVED** [review]

**Comment by bader** [comment]
@GeorgeWeb, we recently switched to opaque pointers by default in the sycl branch. Could you merge the tip of the `sycl` branch your branch to re-test with the recent state of the `sycl` branch, please?

File: clang/test/Driver/sycl-no-gpu-rdc-opaque-pointers.cpp
Comment (bader) [inline clang/test/Driver/sycl-no-gpu-rdc-opaque-pointers.cpp:8]: @mdtoguchi, @sarnex, I think SPIR-V supports opaque pointers already.
I'm not sure I understand what exactly has to be removed. Could you check if we need to make any additional changes to this patch, please?

Comment (sarnex) [reply clang/test/Driver/sycl-no-gpu-rdc-opaque-pointers.cpp:8]: Sorry for late response. Sounds good, thanks.

**Pull request merged**
